### PR TITLE
Background corrections for IntegrateEllipsoids

### LINKED ISSUE #12867: Sigma from the background may be too small.
Check with Xiaoping Wang at TOPAZ for details.

### PR DESCRIPTION
**Pull request by VickieLynch**
Fixes #12867.  Sort background events and remove top 1%.  Also add option, IntegrateIfOnEdge=False to check if integration radius if off edge of detectors and correct background if background radius is off edge.

To test

``` python
Load(Filename='TOPAZ_11052_event.nxs', OutputWorkspace='TOPAZ_11052_event', LoadMonitors=True)
ConvertToMD(InputWorkspace='TOPAZ_11052_event', QDimensions='Q3D', dEAnalysisMode='Elastic', Q3DFrames='Q_sample', LorentzCorrection=True, OutputWorkspace='TOPAZ_11052_md', MinValues='-25,-25,-25', MaxValues='25,25,25', SplitInto='2', SplitThreshold=50, MaxRecursionDepth=13, MinRecursionDepth=7)
FindPeaksMD(InputWorkspace='TOPAZ_11052_md', PeakDistanceThreshold=0.37680000000000002, MaxPeaks=1000, DensityThresholdFactor=1, OutputWorkspace='TOPAZ_11052_peaks')
FindUBUsingFFT(PeaksWorkspace='TOPAZ_11052_peaks', MinD=3, MaxD=15, Tolerance=0.12)
IndexPeaks(PeaksWorkspace='TOPAZ_11052_peaks', Tolerance=0.12)
ShowPossibleCells(PeaksWorkspace='TOPAZ_11052_peaks', BestOnly=False, AllowPermutations=False)
SelectCellOfType(PeaksWorkspace='TOPAZ_11052_peaks', CellType='Monoclinic', Apply=True)
IntegratePeaksMD(InputWorkspace='TOPAZ_11052_md', PeakRadius=0.23999999999999999, BackgroundOuterRadius=0.27000000000000002, PeaksWorkspace='TOPAZ_11052_peaks', OutputWorkspace='sphere', IntegrateIfOnEdge=False)
IntegrateEllipsoids(InputWorkspace='TOPAZ_11052_event', PeaksWorkspace='TOPAZ_11052_peaks', RegionRadius=0.24, OutputWorkspace='ellipsoid')
IntegrateEllipsoids(InputWorkspace='TOPAZ_11052_event', PeaksWorkspace='TOPAZ_11052_peaks', RegionRadius=0.24, OutputWorkspace='ellipsoid_NotEdge',IntegrateIfOnEdge=False)
```

### CODE REVIEW

**Comment by OwenArnold** [comment]
If you haven't yet added anything yet to the release notes, just a reminder to do that.

File: Code/Mantid/Framework/MDAlgorithms/src/IntegrateEllipsoids.cpp [hunk 1]
```diff
@@ -275,6 +276,11 @@ void IntegrateEllipsoids::init() {
   declareProperty(
       "IntegrateInHKL", false,
       "If true, integrate in HKL space not Q space.");
+
+  declareProperty(
+      "IntegrateIfOnEdge", true,
+      "Only warning if all of peak outer radius is not on detector (default).\n"
```
Comment (OwenArnold) [inline Code/Mantid/Framework/MDAlgorithms/src/IntegrateEllipsoids.cpp]: I think the property description could be written a little clearer.

File: Code/Mantid/Framework/MDAlgorithms/src/IntegrateEllipsoids.cpp [hunk 2]
```diff
@@ -399,9 +422,20 @@ void IntegrateEllipsoids::exec() {
       peak_q = peaks[i].getQLabFrame();
       std::vector<double> axes_radii;
       Mantid::Geometry::PeakShape_const_sptr shape =
-          integrator.ellipseIntegrateEvents(
+          integrator.ellipseIntegrateEvents(E1Vec,
               peak_q, specify_size, peak_radius, back_inner_radius,
               back_outer_radius, axes_radii, inti, sigi);
+      // Do not integrate if sphere is off edge of detector
+
+      /*if (!detectorQ(peak_q, axes_radii)) {
```
Comment (OwenArnold) [inline Code/Mantid/Framework/MDAlgorithms/src/IntegrateEllipsoids.cpp]: Better to remove dead code rather than leave it commented out.

**Comment by OwenArnold** [comment]
It didn't look like the rst had been modified. I'd like to know where the 1% aspect came from, and it would be useful for other scientists to know what processing is happing with the peak background. Could you add a section on this to the rst?

**Comment by ricleal** [comment]
I'm assuming that the intensity of peaks on the edge are zeroed. If so it looks correct to me.

**Comment by ricleal** [comment]
I'm assuming that the intensity of peaks on the edge are zeroed. If so it looks correct to me.

**Pull request merged**